subsys/settings: Allow to register custom backends in settings subsystem

fnde-ot opened the pull request:
This is a rework of the settings subsystem based on closed PR #12998 

Major changes are:
- Expose settings backend API to enable custom backend support.
- Add a new CONFIG_SETTINGS_CUSTOM backend to allow registering a custom
backend.
- Change api of the handlers h_set() routines to allow for
backend-specific read callbacks.
- Provide a customizable settings_backend_init() routine for custom
backends.
- Move runtime settings support to be its own backend.

Signed-off-by: François Delawarde (FNDE) <fnde@oticon.com>

Vudentz reviewed (changes requested):
It looks like there are a lot of unrelated changes to bluetooth, mostly code style changes which probably should be either left alone or submited in a different patch.

fnde-ot commented:
> It looks like there are a lot of unrelated changes to bluetooth, mostly code style changes which probably should be either left alone or submited in a different patch.

@Vudentz, Sorry about that, I followed the contribution guidelines and ran uncrustify and checkpatch on all files in the diff, and the previous contributor might have skipped that step. I will try to move those unrelated code style changes into another commit.

fnde-ot committed:
subsys/settings: Allow to register custom backends in settings subsystem

Major changes are:
- Expose settings backend API to enable custom backend support.
- Add a new CONFIG_SETTINGS_CUSTOM backend to allow registering a custom
backend.
- Change api of the handlers h_set() routines to allow for
backend-specific read callbacks.
- Provide a customizable settings_backend_init() routine for custom
backends.
- Move runtime settings support to be its own backend.

Signed-off-by: François Delawarde <fnde@oticon.com>

dbkinder commented on doc/reference/storage/settings/settings.rst:
```diff
@@ -40,11 +40,37 @@ These are registered using a call to ``settings_register()``.
+***********************
+
+Zephyr has two existing backend storages which can be a Flash Circular Buffer
+(CONFIG_SETTINGS_FCB) or a file in the filesystem (CONFIG_SETTINGS_FS).
```
Link to the option documentation:
```suggestion
(:option:`CONFIG_SETTINGS_FCB`) or a file in the filesystem (:option:`CONFIG_SETTINGS_FS`).
```

fnde-ot replied:
Fixed, thanks!

dbkinder commented on doc/reference/storage/settings/settings.rst:
```diff
@@ -152,6 +178,37 @@ up from where it was before restart.
+**************************************
+
+This is a simple example showing how to register a simple custom backend
+handler (CONFIG_SETTINGS_CUSTOM).
```
```suggestion
handler (:option:`CONFIG_SETTINGS_CUSTOM`).
```

fnde-ot replied:
Fixed, thanks!

dbkinder reviewed (changes requested):
Looks good with a couple of edits

dbkinder reviewed (approved):
+1 for doc changes, thanks.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
+			len = read_cb(
+			 cb_arg,
+			 &reset_counter,
+			 sizeof(reset_counter));
```
This looks like very strange style - surely you can fit at least some of these on the same line?

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
+			len = read_cb(
+			 cb_arg,
+			 &gen_def_trans_time_srv_user_data.tt,
+			 sizeof(gen_def_trans_time_srv_user_data.tt));
```
Same here, though I realise the old code in this case had the strange style as well.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
+			len = read_cb(
+			 cb_arg,
+			 &gen_power_onoff_srv_user_data.onpowerup,
+			 sizeof(gen_power_onoff_srv_user_data.onpowerup));
```
And here - put arguments which fit on the same line on the same line

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
+			len = read_cb(
+			 cb_arg,
+			 &light_ctl_srv_user_data.lightness_temp_def,
+			 sizeof(light_ctl_srv_user_data.lightness_temp_def));
```
and here

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
+			len = read_cb(
+			 cb_arg,
+			 &light_ctl_srv_user_data.lightness_temp_last,
+			 sizeof(light_ctl_srv_user_data.lightness_temp_last));
```
and here

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
+			len = read_cb(
+			 cb_arg,
+			 &light_lightness_srv_user_data.lightness_range,
+			 sizeof(light_lightness_srv_user_data.lightness_range));
```
and here

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c:
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
+			len = read_cb(
+			 cb_arg,
+			 &light_ctl_srv_user_data.temperature_range,
+			 sizeof(light_ctl_srv_user_data.temperature_range));
```
and here

jhedberg reviewed (changes requested):
Looks generally fine to me, except for some coding style issues in the onoff_level_lightning_vnd_app

carlescufi commented:
@nvlsianpu can you take a look at this one?

fnde-ot commented on doc/reference/storage/settings/settings.rst:
```diff
@@ -40,11 +40,38 @@ These are registered using a call to ``settings_register()``.
+Backends
+********
+
+Backends are meant to load and save data from/to persistent storage, and
```
True that!

fnde-ot commented on include/settings/settings.h:
```diff
@@ -151,88 +153,58 @@ int settings_save_one(const char *name, void *value, size_t val_len);
 	const struct settings_store_itf *cs_itf;
 };
 
+struct settings_store_itf {
```
Will do.

fnde-ot commented on include/settings/settings.h:
```diff
@@ -151,88 +153,58 @@ int settings_save_one(const char *name, void *value, size_t val_len);
 };
 
+struct settings_store_itf {
+	int (*csi_load)(struct settings_store *cs);
```
I agree, so do the export and get functions in module handlers, but I would rather that being done in another PR if it is okay.

nvlsianpu replied:
Ok, maybe we can do that in coming PRs (will see)

fnde-ot commented on subsys/settings/Kconfig:
```diff
@@ -35,7 +38,7 @@ config SETTINGS_USE_BASE64
 choice
 	prompt "Storage back-end"
-	default SETTINGS_FCB if FCB
+	default SETTINGS_NONE
```
It makes a lot of qemu sanitycheck tests fail. For example samples/bluetooth/peripheral which selects FCB even though it is not supported for qemu_x86. That change makes it fall back to the default SETTINGS_NONE.

fnde-ot replied:
Resolved by putting both lines in, so it falls back to SETTINGS_NONE if there FCB is not enabled.

jukkar commented:
Merging this one as requested by @aescolar as most of the needed people are in wrong timezone.
